tensorflow/java/src/gen/java/org/tensorflow/processor/operator/OperatorProcessor.java

```diff
@@ -0,0 +1,115 @@
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import javax.annotation.processing.AbstractProcessor;
+import javax.annotation.processing.Filer;
+import javax.annotation.processing.Messager;
+import javax.annotation.processing.ProcessingEnvironment;
+import javax.annotation.processing.RoundEnvironment;
+import javax.lang.model.SourceVersion;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.TypeElement;
+import javax.tools.Diagnostic.Kind;
+
+public final class OperatorProcessor extends AbstractProcessor {
+
+  @Override
+  public SourceVersion getSupportedSourceVersion() {
+    return SourceVersion.latestSupported();
+  }
+
+  @Override
+  public synchronized void init(ProcessingEnvironment processingEnv) {
+    super.init(processingEnv);
+    messager = processingEnv.getMessager();
+    filer = processingEnv.getFiler();
+  }
+
+  @Override
+  public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
+    // Nothing needs to be done at the end of all rounds.
+    if (roundEnv.processingOver()) {
+      return false;
+    }
+
+    // Collect all classes tagged with our annotation.
+    Set<TypeElement> opClasses = new HashSet<TypeElement>();
+    for (TypeElement annotation : annotations) {
+      if (!collectOpClasses(roundEnv, opClasses, annotation)) {
+        return true;
+      }
+    }
+
+    // Nothing to do when there are no tagged classes.
+    if (opClasses.isEmpty()) {
+      return false;
+    }
+
+    // TODO:(kbsriram) validate operator classes and generate Op API.
+    writeApi();
+    hasRun = true;
+    return true;
+  }
+
+  @Override
+  public Set<String> getSupportedAnnotationTypes() {
+    return Collections.singleton(String.format("%s.annotation.Operator", OP_PACKAGE));
+  }
+
+  private void writeApi() {
+    // Generate an empty class for now and get the build working correctly. This will be changed to
+    // generate the actual API once we've done with build-related changes.
+    // TODO:(kbsriram)
+    try (PrintWriter writer =
+        new PrintWriter(filer.createSourceFile(String.format("%s.Ops", OP_PACKAGE)).openWriter())) {
+      writer.println(String.format("package %s;", OP_PACKAGE));
+      writer.println("public class Ops{}");
+    } catch (IOException e) {
+      error(null, "Unexpected failure generating API: %s", e.getMessage());
+    }
+  }
+
+  private boolean collectOpClasses(
+      RoundEnvironment roundEnv, Set<TypeElement> opClasses, TypeElement annotation) {
+    boolean result = true;
+    for (Element e : roundEnv.getElementsAnnotatedWith(annotation)) {
+      // This processor has to aggregate all op classes in one round, as it generates a single Ops
+      // API class which cannot be modified once generated. If we find an annotation after we've
+      // generated our code, flag the location of each such class.
+      if (hasRun) {
```

Comment:
Instead of doing this in the loop, what about the following:

- Make this method:

```java
private boolean collectOpClasses(Set<? extends Element> elements, Set<TypeElement> opClasses) {
  for (Element e : elements) {
    ...
  }
}
```

- Make the caller (`process`) do something like:

```java
assert annotations.size() == 0;
Set<? extends Element> annotated = roundEnv.getElementsAnnotatedWith(annotation);
if (annotated.size() == 0) {
  return true;
}
if (annotated.size() > 0 && hasRun) {
  // print error
  return false;
}
if (!collectOpClasses(annotated, opClasses)) {
  return false;
}
...
```

It's essentially the same, but I think it makes the expected constraints clearer to follow. Though, it's possible you disagree :)

Reply:
Done, combined with the previous suggestion.